analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/HardcodedBytesRuleBase.cs

```diff
@@ -40,12 +40,22 @@ protected ProgramState ProcessArrayCreation(ProgramState state, IArrayCreationOp
     }
 
     // array[42] = ...
-    protected ProgramState ProcessArrayElementReference(ProgramState state, IArrayElementReferenceOperationWrapper arrayElementReference) =>
-        (arrayElementReference.IsAssignmentTarget() || arrayElementReference.IsCompoundAssignmentTarget())
+    protected ProgramState ProcessArrayElementReference(ProgramState state, IArrayElementReferenceOperationWrapper arrayElementReference)
+    {
+        return (IsAssignedToNonConstant() || IsCompoundAssignedToNonConstant())
         && arrayElementReference.ArrayReference.TrackedSymbol(state) is { } array
             ? state.SetSymbolConstraint(array, NotHardcoded)
             : state;
 
+        bool IsAssignedToNonConstant() =>
+            arrayElementReference.IsAssignmentTarget()
+            && !ISimpleAssignmentOperationWrapper.FromOperation(arrayElementReference.ToSonar().Parent).Value.ConstantValue.HasValue;
+
+        bool IsCompoundAssignedToNonConstant() =>
+            arrayElementReference.IsCompoundAssignmentTarget()
+            && !ICompoundAssignmentOperationWrapper.FromOperation(arrayElementReference.ToSonar().Parent).Value.ConstantValue.HasValue;
+    }
```

Comment:
Before I was a bit lazy and did not check the RHS value, now this check is more precise.
I think this is a very rare case, but since you added a test for it, I added support for it :)